lib/puppet/agent.rb

```diff
@@ -60,8 +61,22 @@ def run(client_options = {})
               end
             end
           rescue Puppet::LockError
-            Puppet.notice _("Run of %{client_class} already in progress; skipping  (%{lockfile_path} exists)") % { client_class: client_class, lockfile_path: lockfile_path }
-            return
+            wait_for_lock_deadline ||= Time.now.to_i + Puppet[:maxwaitforlock]
+
+            if Puppet[:waitforlock]  < 1
+              Puppet.notice _("Run of %{client_class} already in progress; skipping  (%{lockfile_path} exists)") % { client_class: client_class, lockfile_path: lockfile_path }
+              return 1
+            elsif Time.now.to_i > wait_for_lock_deadline
+              Puppet.notice _("Exiting now because the maxwaitforlock timeout has been exceeded.")
+              return 1
+            else
+              while File.file?(lockfile_path) && Time.now.to_i <= wait_for_lock_deadline
```

Comment:
If the process holding the lock doesn't cleanup correctly (and delete the lockfile), then we'll wait until the deadline occurs and then retry. The retry might acquire the lock, but that means we've spent on average 30 seconds (half the maxwaitforlock) sleeping when we could have acquired the lock. Maybe just retry the lock each time?

```ruby
if Puppet[:waitforlock]  < 1
  ...
elsif Time.now.to_i > wait_for_lock_deadline
  ...
else # Time.now.to_i <= wait_for_lock_deadline
  Puppet.info _("Another puppet instance is already running; --waitforlock flag used, waiting for running instance to finish.")
  Puppet.info _("Will try again in %{time} seconds.") % {time: Puppet[:waitforlock]}
  sleep Puppet[:waitforlock]
  retry
end
```